### ISSUE TITLE
Constant assignment in non-static methods

### ISSUE THREAD

**Issue by felixvf** [mirah/mirah#401]
Consider this Mirah code:

```mirah
class Foo
	def foo
		CONSTANT1 = 1
		puts "a"
	end
	
	def self.bar
		CONSTANT2 = 2
		puts "b"
	end
	
	class << self
		def baz:void
			CONSTANT3 = 3
			puts "c"
		end
	end
	
	CONSTANT4 = 4
	
	def self.p
		puts CONSTANT1
		puts CONSTANT2
		puts CONSTANT3
		puts CONSTANT4
	end
end

Foo.p
Foo.baz
Foo.p
```

When compiling and executing, what should be the result?

1.  

        0
        0
        0
        4
        c
        0
        0
        3
        4
2. 

        1
        2
        3
        4
        c
        1
        2
        3
        4

3. Compilation error

It turns out that option 1. is what currently is implemented. This effectively means that constants are not constants but class variables, such as ```@@CONSTANT1```. 

To make constants actually constant, we should decide for either 2. (allow constant assignment in methods, but the actual assignments happens in a static initializer) or 3. (prohibit constant assignment in methods).

As 2. provides only marginal benefit (more locality of where code is defined to where it is used) and substantial cause for confusion (the right-hand-side of the constant assignment is not evaluated during method invocation, but it looks like being evaluated during method invocation), I'd vote for 3.

**Comment by uujava** [comment]
My vote would be to make upper case constants to always have <public static final> behaviour.
Allow to define and initialize them in self.initialize or implicit self initialize:
```
class Foo
   CONSTANT1 = 1
   def self.initialize
      CONSTANT2 = 2
   end
   class << self
        def initialize:void
            CONSTANT3 = 3
           # CONSTANT2 = 2  -> compile time error at clean step
        end
    end
end
```

I see two options for an upper cased @@XXXX fields 
```
 @@CONSTANT
```
Option#1: always infer @@ fields as private static nonfinal no matter what case are they. Let public nonfinal access to be implemented via optional macro:
```
class Foo
   self_attr_accessor NON_FINAL:int
end
```

that generates:
```
class Foo
   self.NON_FINAL:int
      @@NON_FINAL
   end

   self.NON_FINAL_set(value:int):void
      @@NON_FINAL = value
   end
end
```

Option#2: infer @@ fields as private static nonfinal by default.  Implement modifiers (macro/or parser) to widen access:
```
class Foo   
   public @@CONSTANT1 = 1 # public nonfinal
   private final @@serialVersionUID = 1 # private final
   private volatile @@
   public final @@CONSTANT3 = 1 # effectively same as just CONSTANT3. So we could remove  'upper cased constant' support entierly 
   def self.initialize
      @@CONSTANT1 = 1 # ok
      @@CONSTANT3 = 1 # 
   end
end
```

I'd vote for Option#2

**Comment by felixvf** [comment]
One aspect of your option#2 already works out of the box:

```mirah
class Foo
	class << self
		attr_accessor NON_FINAL:int
	end
end
```

This declares a private static non-final field ```NON_FINAL``` and public static accessor methods ```NON_FINAL():int``` as well as ```NON_FINAL_set(value:int)```.

Modifiers on ConstantAssignment are currently missing, as far as I know.

So you suggest to keep constants private by default? Does this mean that you do not suggest to make constants public by default anymore, as you did [here](https://github.com/mirah/mirah/issues/400#issuecomment-167204447)?

**Comment by uujava** [comment]
good catch about class level attr accessor. In https://github.com/mirah/mirah/issues/400#issuecomment-167204447 I meant the least effort option. Option#2 I guess would take more efforts to imlement, but in my opinion more consistent.
